kotlin-native/NEW_MM.md

```diff
@@ -122,65 +116,52 @@ kotlin.targets.withType(KotlinNativeTarget::class.java) {
-  worker manually processes events with `Worker.processQueue`.
+Currently, the logs are only printed to stderr. _Note that the exact contents of the logs will change._
+
+The list of known performance issues:
+
+* Since the collector is single-threaded stop-the-world, the pause time of every thread linearly depends on the number of objects in the heap. The more objects that are kept alive, the longer the pauses are. Long pauses on the main thread can result in laggy UI event handling. Both the pause time and the number of objects in the heap are printed to the logs for each GC cycle.
+* Being stop-the-world also means that all threads with Kotlin/Native runtime active on them need to synchronize simultaneously for the collection to begin. This also affects the pause time.
+* There is a complicated relationship between Swift/ObjC objects and their Kotlin/Native counterparts, which causes Swift/ObjC objects to linger longer than necessary. It means that their Kotlin/Native counterparts are kept in the heap longer, contributing to the slower collection time. This typically doesn't happen, but in some corner cases, for example, when a long loop creates several temporary objects that cross the Swift/ObjC interop boundary on each iteration (for example, calling a Kotlin callback from a loop in Swift or vice versa).
+  In the logs, there are several stable refs in the root set. If this number keeps growing, it may indicate that the Swift/ObjC objects are not being freed when they should. Try putting `autoreleasepool` around loop bodies (both in Swift/ObjC and Kotlin) that do interop calls.
+* Our GC triggers do not adapt to the workload: collections may be requested far more frequently than necessary, which means that GC time may dominate useful application run time and pause the threads more frequently than needed.
+  This manifests in time between cycles being close (or even less) than the pause time. Both of these numbers are printed to the logs.
+  Try increasing `kotlin.native.internal.GC.threshold` and `kotlin.native.internal.GC.thresholdAllocations` to force GC to happen less often. Note that the exact meaning of `threshold` and `thresholdAllocations` may change in the future.
+* Freezing is currently implemented suboptimally: internally, a separate memory allocation may occur for each frozen object (this recursively includes the object subgraph), which puts unnecessary pressure on the heap.
+* Unterminated `Worker`s and unconsumed `Future`s have objects pinned to the heap, contributing to the pause time. Like Swift/ObjC interop, this also manifests in a growing number of stable refs in the root set.
+  To mitigate:
+    * Look for `Worker.execute` being called with the resulting `Future` and never consumed (via `Future.consume` or `Future.result`) and make sure to either consume the `Future` or replace calls with `Worker.executeAfter` instead.
```

Comment:
> Look for `Worker.execute` being called with the resulting `Future` and never consumed

`Worker.execute` does not take a `Future`, it returns it.

Reply:
Rephrased this part, hope it got clearer

Reply:
It's probably a bit confusing still:

>  Worker.execute that returns a Future being called but never consumed

Seems to imply that the `Future` is called.

`Worker.execute()` is called, it's return value is an object of type `Future`. If the user never calls `.consume()` or `.value` on this return value, then bad things are going to happen performance-wise. What we suggest is to make sure every `Future` returned from `Worker.execute()` is consumed at some point, or the `Worker.execute()` be replaced with `Worker.executeAfter()`, which returns `Unit`, and doesn't have this problem.

Reply:
Thanks for the detailed explanation! I tried a new version here